application/src/test/java/org/thingsboard/server/service/subscription/TbEntityLocalSubsInfoTest.java

```diff
@@ -0,0 +1,177 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.thingsboard.server.service.subscription;
+
+import org.junit.jupiter.api.Test;
+import org.thingsboard.server.common.data.id.DeviceId;
+import org.thingsboard.server.common.data.id.EntityId;
+import org.thingsboard.server.common.data.id.TenantId;
+import org.thingsboard.server.common.data.plugin.ComponentLifecycleEvent;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+class TbEntityLocalSubsInfoTest {
+
+    @Test
+    void addTest() {
+        Set<TbAttributeSubscription> expectedSubs = new HashSet<>();
+        TbEntityLocalSubsInfo subsInfo = createSubsInfo();
+        TenantId tenantId = subsInfo.getTenantId();
+        EntityId entityId = subsInfo.getEntityId();
+        TbAttributeSubscription attrSubscription1 = TbAttributeSubscription.builder()
+                .sessionId("session1")
+                .tenantId(tenantId)
+                .entityId(entityId)
+                .keyStates(Map.of("key1", 1L, "key2", 2L))
+                .build();
+        expectedSubs.add(attrSubscription1);
+        TbEntitySubEvent created = subsInfo.add(attrSubscription1);
+        assertFalse(subsInfo.isEmpty());
+        assertNotNull(created);
+        assertEquals(expectedSubs, subsInfo.getSubs());
+        checkEvent(created, expectedSubs, ComponentLifecycleEvent.CREATED);
+
+        assertNull(subsInfo.add(attrSubscription1));
+
+        TbAttributeSubscription attrSubscription2 = TbAttributeSubscription.builder()
+                .sessionId("session2")
+                .tenantId(tenantId)
+                .entityId(entityId)
+                .keyStates(Map.of("key3", 3L, "key4", 4L))
+                .build();
+        expectedSubs.add(attrSubscription2);
+        TbEntitySubEvent updated = subsInfo.add(attrSubscription2);
+        assertNotNull(updated);
+
+        assertEquals(expectedSubs, subsInfo.getSubs());
+        checkEvent(updated, expectedSubs, ComponentLifecycleEvent.UPDATED);
+    }
+
+    @Test
+    void removeTest() {
```

Comment:
Add one more remove and removeAll test group where you remove only one subscription.